apollo-runtime/src/main/java/com/apollographql/apollo/internal/response/RealResponseWriter.kt

```diff
@@ -0,0 +1,224 @@
+import com.apollographql.apollo.api.internal.ResponseWriter
+import java.math.BigDecimal
+import java.util.ArrayList
+import java.util.LinkedHashMap
+
+class RealResponseWriter(private val operationVariables: Operation.Variables, private val scalarTypeAdapters: ScalarTypeAdapters) : ResponseWriter {
+  val buffer: MutableMap<String, FieldDescriptor> = LinkedHashMap()
+  override fun writeString(field: ResponseField, value: String?) {
+    writeScalarFieldValue(field, value)
+  }
+
+  override fun writeInt(field: ResponseField, value: Int?) {
+    writeScalarFieldValue(field, if (value != null) BigDecimal.valueOf(value.toLong()) else null)
+  }
+
+  override fun writeLong(field: ResponseField, value: Long?) {
+    writeScalarFieldValue(field, if (value != null) BigDecimal.valueOf(value) else null)
+  }
+
+  override fun writeDouble(field: ResponseField, value: Double?) {
+    writeScalarFieldValue(field, if (value != null) BigDecimal.valueOf(value) else null)
+  }
+
+  override fun writeBoolean(field: ResponseField, value: Boolean?) {
+    writeScalarFieldValue(field, value)
+  }
+
+  override fun writeCustom(field: ResponseField.CustomTypeField, value: Any?) {
+    val typeAdapter = scalarTypeAdapters.adapterFor<Any>(field.scalarType)
+    writeScalarFieldValue(field, if (value != null) typeAdapter.encode(value).value else null)
+  }
+
+  override fun writeObject(field: ResponseField, marshaller: ResponseFieldMarshaller?) {
+    checkFieldValue(field, marshaller)
+    if (marshaller == null) {
+      buffer[field.responseName] = FieldDescriptor(field, null)
+      return
+    }
+    val nestedResponseWriter = RealResponseWriter(operationVariables, scalarTypeAdapters)
+    marshaller.marshal(nestedResponseWriter)
+    buffer[field.responseName] = FieldDescriptor(field, nestedResponseWriter.buffer)
+  }
+
+  override fun writeFragment(marshaller: ResponseFieldMarshaller?) {
+    marshaller?.marshal(this)
+  }
+
+  override fun <T> writeList(field: ResponseField, values: List<T>?, listWriter: ResponseWriter.ListWriter<T>) {
+    checkFieldValue(field, values)
+    if (values == null) {
+      buffer[field.responseName] = FieldDescriptor(field, null)
+      return
+    }
+    val accumulated = ArrayList<Any?>()
+    listWriter.write(values, ListItemWriter(operationVariables, scalarTypeAdapters, accumulated))
+    buffer[field.responseName] = FieldDescriptor(field, accumulated)
+  }
+
+  fun resolveFields(delegate: ResolveDelegate<Map<String, Any>?>) {
+    resolveFields(operationVariables, delegate, buffer)
+  }
+
+  private fun writeScalarFieldValue(field: ResponseField, value: Any?) {
+    checkFieldValue(field, value)
+    buffer[field.responseName] = FieldDescriptor(field, value)
+  }
+
+  private fun rawFieldValues(buffer: Map<String, FieldDescriptor>?): Map<String, Any?> {
+    val fieldValues: MutableMap<String, Any?> = LinkedHashMap()
+    for ((fieldResponseName, value) in buffer!!) {
```

Comment:
If this is the case, buffer above shouldn't be nullable

Reply:
I could remove this `!!` and a few other ones. There are still a few in the file that I'm not sure we can remove easily/safely.